src/securemul.rs

```diff
@@ -241,8 +260,63 @@ mod tests {
         assert_eq!(Fp31::from(24_u128), validate_and_reconstruct(result_shares));
     }
 
+    /// This test ensures that many secure multiplications can run concurrently as long as
+    /// they all have unique id associated with it. Basically it validates
+    /// `TestHelper`'s ability to distinguish messages of the same type sent towards helpers
+    /// executing multiple same type protocols
+    #[tokio::test]
+    #[allow(clippy::cast_possible_truncation)]
+    pub async fn concurrent_mul() {
+        let ring = helpers::ring::mock::make_three();
+        let participants = crate::prss::test::make_three();
+        let context = make_context(&ring, &participants);
+        let mut rand = StepRng::new(1, 1);
+        let a = share(Fp31::from(4_u128), &mut rand);
+        let b = share(Fp31::from(3_u128), &mut rand);
+
+        let mut multiplications = Vec::new();
+        for i in 1..10_u128 {
+            // there is something weird going on the compiler's side. I don't see why we need
+            // to use async move as `i` is Copy + Clone, but compiler complains about it not living
+            // long enough
+            let ctx = &context;
+            let f = async move {
+                let h1_future = SecureMul {
+                    step: TestStep::Mul1(i as u8),
```

Comment:
Please don't use the index as an argument to the step.  That will mean that you have separation at the step AND index layer, but based on the same value.  It's not redundant, but aligned, which isn't what we are looking for.

What you have here is a case where there are three helpers acting, so each should just use TestStep::Mul(0) or some other fixed step identifier.  They are each operating on the same data, so they will have the same index too.


I don't know if this test will pass if you make this change, because I'm not seeing anywhere that guarantees that the test infrastructure maintains order based on index.  What you have done might be hiding that as a problem.

Reply:
I agree that it is better to semantically separate index and step. However, we seem to require "dynamic" steps (where step is essentially composed of Step name + index, for example in share conversion we use `Mul(bit_id)` . That is what I tried to reproduce here to make sure we would be able to do it with this change.

So I am trying to understand whether just logically representing share "bit" `i` and `index` as independent counters would suffice or there is a bigger issue that I don't see

Reply:
Yeah, I think that what you want is a bit number and a record number.  You pass the bit number to `TestStep::Mul1(bit_number)` and the record number (or index) goes in `index`.